95 add multi version support

popescu-v opened the pull request:
Add support for two versions:
- current version, accessible from the base URL of the site
- other version x.y.z[-pre], accessible from the base URL/x.y.z[-pre]

Note: the deployment part should be tested as soon as we have the 10.3.2 / 10.3.2.0 Khiops core / Python library release combo.

popescu-v committed:
Include the Python API doc in the web site build

This has the effect of including the Python API static site into the
global web site.

This works in deployment as well, because the Python API static files
are not committed, nor are they known to the Git repository, hence they
are being included in the build _and_deployed.
See https://www.mkdocs.org/user-guide/deploying-your-docs/.

popescu-v committed:
Set Python API docs URI to the local path (as set in the CI/CD)

popescu-v committed:
Store site build as artifact for off-line inspection

popescu-v committed:
Don't deploy to GH pages by default; require specific user action

popescu-v committed:
Hide release tag from the repository information

Otherwise, only the latest release is displayed.

Thus, this doesn't work correctly for:
- beta releases (the latest non-beta is displayed)
- simultaneous v10 and v11 releases (the latest v11 would be displayed).

The behavior which consists of only displaying the latest release tag is
hard-coded and cannot be easily overridden.

popescu-v committed:
Make mkdocs extra variables settable from the environment

Thus, they can be configured from the CI/CD, without requiring
supplemental commits for their update.

popescu-v committed:
Add two-version support: current and other version

TODO: Replace gh-deploy with the deploy-pages GitHub action.

popescu-v committed:
Stop caching mkdocs pages, as this is not strictly necessary

popescu-v committed:
Configure Git credentials right before the deployment, where necessary

popescu-v committed:
Fix environment file access: first checkout the repo to access the file

popescu-v committed:
Don't pollute the environment with build-specific environment variables

Thus, they remain local to the _current iteration_ of the doc build
action.

popescu-v committed:
Only build other version if explicitly requested to do so

popescu-v committed:
Checkout doc repository to the current branch after the action

Thus, we ensure consistent post-action steps in the CI.

popescu-v committed:
Deploy previously-built site rather than rebuilding and deploying it

Thusly, the multi-version site ca be deployed.

popescu-v committed:
Update other version site reference

folmos-at-orange commented on overrides/assets/stylesheets/main.css:
```diff
@@ -138,4 +138,9 @@ footer .mention a:hover {
   content: "\e944";
 }
-
+.md-source__fact--version {
+  display: none;
+}
+.md-source__fact:nth-child(1n+2)::before {
+  margin-left: 0;
+}
```
Maybe a comment similar to that of the commit? For me this is dark magic.

popescu-v replied:
OK, I'll add a comment here.

popescu-v replied:
Done.

folmos-at-orange commented on .github/workflows/ci.yml:
```diff
@@ -3,40 +3,86 @@ name: Website
+          pip install ghp-import
+
+          # Push built site directory contents to GH pages
+          ghp-import -m "Deployed ${GITHUB_SHA:0:7}" --push --force --no-jeckyll ./site/
```
Should'nt be `--no-jekyll` ? [ref](https://github.com/c-w/ghp-import?tab=readme-ov-file#usage)

popescu-v replied:
Right! This step has not been tested (yet), because we need to have the 10.3.2[.0] to do that.

popescu-v replied:
Done.

folmos-at-orange reviewed (approved):
If possible add a section in the README with a high-level description of how the versioning works.
Since we are not using standard versioning, there is a lot done with CI scripts, so it may be useful to have a good mental model for future modifications.

One general question ... we aren't pushing the whole site to the `gh-pages` branch at each deployment ? (maybe I missed the code doing that)

popescu-v commented:
> If possible add a section in the README with a high-level description of how the versioning works. Since we are not using standard versioning, there is a lot done with CI scripts, so it may be useful to have a good mental model for future modifications.
> 
> One general question ... we aren't pushing the whole site to the `gh-pages` branch at each deployment ? (maybe I missed the code doing that)

The push to the `gh-pages` branch is done in the step which calls into the `ghp-import` command.

lucaurelien reviewed (approved):
No issues from a functional standpoint.

popescu-v committed:
Update the README

- separate (local) development / production modes
- document the new CI which supports the production mode

popescu-v commented:
> If possible add a section in the README with a high-level description of how the versioning works. Since we are not using standard versioning, there is a lot done with CI scripts, so it may be useful to have a good mental model for future modifications.

Done: rewrote most of the README to reflect the new workflow.

folmos-at-orange commented on README.md:
````diff
@@ -1,37 +1,76 @@
-```
+The current CI/CD supports building and deploying _two_ versions of the Khiops
+documentation web site:
 
````
I think this should be kept.

popescu-v replied:
IMHO, using `mkdocs serve` no longer provides a relevant development experience, because it would only render the Khiops website proper, without the Python documentation part and with only a single version, namely the one whose default values are specified in `mkdocs.yml` (unless of course the developer overrides these via the relevant environment variables, but this is tedious to do by hand, for little gain: only a single version would be displayed, without the Python part).

IMHO, the proper way to proceed now is to launch the CI _without_ requiring to deploy, then fetch the generated artifact, unzip it and inspect the web site locally (as documented in the updated README file).

popescu-v replied:
But I can retain the "quick feedback" argument. Hence, I wil try to reinstate this part, with the caveat mentioned in the previous comment.

popescu-v replied:
Done: reinstated the local development documentation in a separate section.

popescu-v committed:
Hide the logo from the Python API docs pages

The Python API docs are displayed as iframes and the logo is already in the
enclosing pages.
